File: 133-clone-graph/clone-graph.js
/**
 * // Definition for a _Node.
 * function _Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {_Node} node
 * @return {_Node}
 */
var cloneGraph = function(node) {
    if (!node) return null;
    const nodeMap = new Map();
    const stack = [node];
    const visited = new Set();
    let head;
    while (stack.length) {
        let curr = stack.pop();
        if (!nodeMap.has(curr)) {
            nodeMap.set(curr, new _Node(curr.val));
        }
        let copy = nodeMap.get(curr);
        visited.add(curr);
        for (let n of curr.neighbors) {
            if (!nodeMap.has(n)) nodeMap.set(n, new _Node(n.val));
            copy.neighbors.push(nodeMap.get(n));
            if (!visited.has(n)) {
                stack.push(n);
                visited.add(n);
            }
        }
        if (!head) {
            head = copy;
        }
    }
    return head;
};

File: 133-clone-graph/clone-graph.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var neighbors: [Node?]
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.neighbors = []
 *     }
 * }
 */

class Solution {
    func cloneGraph(_ root: Node?) -> Node? {
        guard var n = root else { return root }
        var map: [Int: Node] = [:], visited: Set<Int> = Set<Int>(), stack: [Node] = [n]
        while !stack.isEmpty {
            var node: Node = stack.removeLast()
            if map[node.val] == nil { map[node.val] = Node(node.val)}
            var oNode: Node = map[node.val]!
            visited.insert(oNode.val)
            for child in node.neighbors {
                if map[child!.val] == nil { map[child!.val] = Node(child!.val) }
                var oChild: Node = map[child!.val]!
                if !visited.contains(oChild.val) {
                    stack.append(child!)
                    visited.insert(oChild.val)
                }
                oNode.neighbors.append(oChild)
            }
        }
        return map[n.val]
    }
}